sdks/python/apache_beam/runners/portability/stager.py

```diff
@@ -0,0 +1,556 @@
+
+If --requirements_file is present then the file specified by the option will be
+staged in the staging location.  When a worker starts it will check for the
+presence of this file and will run 'pip install -r requirements.txt'. A
+requirements file can be easily generated by running 'pip freeze -r
+requirements.txt'. The reason a runner does not run this automatically is
+because quite often only a small fraction of the dependencies present in a
+requirements.txt file are actually needed for remote execution and therefore a
+one-time manual trimming is desirable.
+
+TODO(silviuc): Staged files should have a job specific prefix.
+To prevent several jobs in the same project stomping on each other due to a
+shared staging location.
+
+TODO(silviuc): Should we allow several setup packages?
+TODO(silviuc): We should allow customizing the exact command for setup build.
+"""
+
+import glob
+import logging
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+
+import pkg_resources
+
+from apache_beam.internal import pickler
+from apache_beam.io.filesystems import FileSystems
+from apache_beam.options.pipeline_options import SetupOptions
+# TODO(angoenka): Remove reference to dataflow internal names
+from apache_beam.runners.dataflow.internal import names
+from apache_beam.utils import processes
+
+# All constants are for internal use only; no backwards-compatibility
+# guarantees.
+
+# Standard file names used for staging files.
+WORKFLOW_TARBALL_FILE = 'workflow.tar.gz'
+REQUIREMENTS_FILE = 'requirements.txt'
+EXTRA_PACKAGES_FILE = 'extra_packages.txt'
+
+# Package names for distributions
+BEAM_PACKAGE_NAME = 'apache-beam'
+
+
+class FileHandler(object):
+
+  def file_copy(self, from_path, to_path):
+    """Copies a local file to a remote location or vice versa."""
+    logging.info('File copy from %s to %s.', from_path, to_path)
+
+    if not os.path.isdir(os.path.dirname(to_path)):
+      logging.info(
+          'Created folder (since we have not done yet, and any errors '
+          'will follow): %s ', os.path.dirname(to_path))
+      os.mkdir(os.path.dirname(to_path))
+    shutil.copyfile(from_path, to_path)
+
+  def file_download(self, from_url, to_path):
+    """Downloads a file over http/https from a or copy them from a remote
+    location."""
+    if from_url.startswith('http://') or from_url.startswith('https://'):
+      # TODO(silviuc): We should cache downloads so we do not do it for every
+      # job.
+      try:
+        # We check if the file is actually there because wget returns a file
+        # even for a 404 response (file will contain the contents of the 404
+        # response).
+        # TODO(angoenka): Extract and use the filename when downloading file.
+        response, content = __import__('httplib2').Http().request(from_url)
+        if int(response['status']) >= 400:
+          raise RuntimeError(
+              'Beam SDK not found at %s (response: %s)' % (from_url, response))
```

Comment:
<!--new_thread; commit:d2647e7c0c6001a89be4dc0a673e17d5ce0d951a; resolved:0-->
Let's remove assumptions that we are downloading SDK, it can be any file.